### PR TITLE
Transformer end to end passage ranking examples 

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by jobergum** [comment]
Need to update document-ranking instructions as well as they share app package (maybe a mistake in retro)

**Comment by jobergum** [comment]
I think this is in pretty good shape now (except that some of the searcher logic could have been prettier @bratseth )

File: msmarco-ranking/passage-ranking.md [hunk 1]
```diff
@@ -91,31 +96,60 @@ schema passage {
       index: enable-bm25
     }
     
-    field dt type tensor&lt;float&gt;(dt{}, x[32]){
+    field text_token_ids type tensor<float>(d0[128])  {
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:99]:
```suggestion
    field text_token_ids type tensor&lt;float&gt;(d0[128])  {
```

File: msmarco-ranking/passage-ranking.md [hunk 2]
```diff
@@ -91,31 +96,60 @@ schema passage {
...
       indexing: summary | attribute
     }
-
+    
+    field dt type tensor&lt;bfloat16&gt;(dt{}, x[32]){
+      indexing: summary | attribute
+    }
+    
+    field mini_document_embedding type tensor<float>(d0[384]) {
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:107]:
```suggestion
    field mini_document_embedding type tensor&lt;float&gt;(d0[384]) {
```

File: msmarco-ranking/passage-ranking.md [hunk 3]
```diff
@@ -161,78 +196,111 @@ rank-profile colbert inherits bm25 {
...
-See [Ranking with ONNX Models](https://docs.vespa.ai/en/onnx.html) and
-[How we accelerate inference using ONNX runtime](https://blog.vespa.ai/stateful-model-serving-how-we-accelerate-inference-using-onnx-runtime/)
-* The trained weights of our model is published on [Hugginface model hub](https://huggingface.co/vespa-engine/colbert-medium) and 
-we provide instructions on howto export the Transformer model to ONNX format 
+### Query encoders 
 
+We represent both the ColBERT and the dense query encoder in Vespa 
+using Vespa's support for inference and [ranking with ONNX](https://docs.vespa.ai/en/onnx.html) models. 
+In this case we don't rank any documents directly with the ONNX model but we use the support to make a single pass through them
+to obtain the embedding vector for nearest neighbor search and the colbert query tensor representation for re-ranking.
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:275]:
```suggestion
to obtain the embedding vector for nearest neighbor search and the ColBERT query tensor representation for re-ranking. 
```

File: msmarco-ranking/passage-ranking.md [hunk 4]
```diff
@@ -268,16 +336,53 @@ Install python dependencies. There are no run time python dependencies in Vespa,
...
 
-Once we have downloaded and exported the model, we use maven to create the
+In total we use three ONNX models, please follow the instructions to download them and 
+place them in the files directory with the correct file name. The mode files are referenced
+in the *passage* and *query* vespa doucment schemas.
+
+## The sentence encoder model (bi-encoder)
+
+This model is used for dense retrieval using HNSW indexing. The model uses mean pooling 
+and we add a L2 normalization on top so we can use innerproduct distance metric instead
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:354]:
```suggestion
and we add an L2 normalization on top so we can use innerproduct distance metric instead
```

File: msmarco-ranking/passage-ranking.md [hunk 5]
```diff
@@ -268,16 +336,53 @@ Install python dependencies. There are no run time python dependencies in Vespa,
...
+<pre data-test="exec">
+$ curl -L -o src/main/application/files/sentence-msmarco-MiniLM-L-6-v3-quantized.onnx \
+    https://data.vespa.oath.cloud/onnx_models/sentence-msmarco-MiniLM-L-6-v3-quantized.onnx
+</pre>
+
+## The cross encoder model (classification)
+
+This is the final re-ranking stage using full all to all
+interaction between the query and the passage with a classification head on top of the CLS token
+embedding.
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:366]:
```suggestion
embedding:
```

File: msmarco-ranking/passage-ranking.md [hunk 6]
```diff
@@ -268,16 +336,53 @@ Install python dependencies. There are no run time python dependencies in Vespa,
...
+interaction between the query and the passage with a classification head on top of the CLS token
+embedding. 
+
+<pre data-test="exec">
+$ curl -L -o src/main/application/files/ms-marco-MiniLM-L-6-v2-quantized.onnx \
+    https://data.vespa.oath.cloud/onnx_models/ms-marco-MiniLM-L-6-v2-quantized.onnx
+</pre>
+
+## The late interaction model 
+This is the ColMiniLM model which uses late contextualized interaction and Vespa MaxSim tensor expression
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:374]:
```suggestion
This is the ColMiniLM model which uses late contextualized interaction and Vespa MaxSim tensor expression:
```

File: msmarco-ranking/passage-ranking.md [hunk 7]
```diff
@@ -362,96 +472,152 @@ $ curl -s -H Content-Type:application/json --data @sample-feed/query.json \
...
-See [MS Marco Passage Ranking](https://microsoft.github.io/MSMARCO-Passage-Ranking/) for how to submit runs for the leaderboard. 
+
+# Appendix ColBERT 
+
+Model training and offline text to tensor processing
+
+* The *ColBERT* model is trained the instructions from the [ColBERT repository](https://github.com/stanford-futuredata/ColBERT) 
+using the MS Marco Passage training set. The *bert-base-uncased* is replaced with the *MiniLM-L6*. 
+We use cosine similarity (innerproduct as the vectors are unit length normalized).
+* The dimensionality of the token tensor is reduced from 384 (hidden dim) to 32 dimensions by a linear layer.
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:618]:
```suggestion
* The dimensionality of the token tensor is reduced from 384 (hidden dim) to 32 dimensions by a linear layer
```

**Review by kkraune — APPROVED** [review]
Great! just nitpicks as usual. I did not review the code but noticed you should add copyright in some files

**Comment by jobergum** [comment]
Thanks a lot for review @kraune!

**Pull request merged**